waspc/src/Wasp/Generator/Shared.hs

```diff
@@ -0,0 +1,31 @@
+module Wasp.Generator.Shared
+  ( mkTmplFdWithDst,
+    SharedTemplatesDir,
+  )
+where
+
+import qualified Data.Aeson as Aeson
+import StrongPath (Dir, File', Path', Rel, reldir, (</>))
+import Wasp.Generator.Common (ProjectRootDir)
+import Wasp.Generator.FileDraft (FileDraft, createTemplateFileDraft)
+import Wasp.Generator.Templates (TemplatesDir)
+
+data SharedTemplatesDir
```

Comment:
Hm, we now have both `Shared` and `Common`.
Sure, one is a module with common functionalities while the other is a module with functionalities related to the shared templates, but the names are still pretty similar (and possibly confusing).

I'm not sure whether it's better to come up with better names, or think about unifying them (they do share a semantic connection, after all).

If we opt for the second option, we can place `SharedTemplatesDir` and `sharedTemplatesDirInTemplatesDir` in _Generator/Common.hs_, and introduce a few changes to the file draft functions already present in _ServerGenerator.hs_ and _WebAppGenerator.hs_.

I'll use the `WebAppGenerator` as an example. This is what we currently have:
```haskell
-- file: Generator/WebAppGenerator/Common.hs

mkTmplFdWithDst :: Path' (Rel WebAppTemplatesDir) File' -> Path' (Rel WebAppRootDir) File' -> FileDraft
mkTmplFdWithDst src dst = mkTmplFdWithDstAndData src dst Nothing
```
Instead, we could have something like this:

```haskell
-- file: Generator/Common.hs

data SharedTemplatesDir


-- file: Generator/WebAppGenerator/Common.hs

class TemplatesDirForWebApp d

instance TemplatesDirForWebApp SharedTemplatesDir
instance TemplatesDirForWebApp WebAppTemplatesDir

mkTmplFdWithDst :: TemplatesDirForWebApp d => Path' (Rel d) File' -> Path' (Rel WebAppRootDir) File' -> FileDraft
mkTmplFdWithDst src dst = mkTmplFdWithDstAndData src dst Nothing
```
Doing so would allow us to use the same polymorphic function for working with any template file that's applicable to WebApp, be it from _shared/templates_ or _web-app/templates_.

Reply:
This is an interesting idea and I quite like it. The `Shared.hs` module definetly felt a bit to me since it's not really standalone `SharedGenerator`.

I'll try using this new try we learned with creating `instances` to limit our polymorphism 🔥

Reply:
I've created a "generic" function for making FileDrafts with destination and data that accepts one of the two template folders `SharedTemplatesDir` and `WebAppTemplatesDir`. 

Then we export two functions `mkTmplWithDst` and `mkSharedTmplWithDst`, it should be clear when the template is coming from the shared folder.

Reply:
I now see this whole isn't playing out as I had hoped. My original suggestion obviously wouldn't work because of the error this code throws:
```haskell
mkAnyValidTmplFdWithDstAndData ::
  ValidWebAppTemplatesDir d =>
  Path' (Rel d) File' ->
  Path' (Rel WebAppRootDir) File' ->
  Maybe Aeson.Value ->
  FileDraft
mkAnyValidTmplFdWithDstAndData relSrcPath relDstPath tmplData =
  createTemplateFileDraft
    (webAppRootDirInProjectRootDir </> relDstPath)
    (webAppTemplatesDirInTemplatesDir </> relSrcPath)
    tmplData
``` 
So, unfortunately, it doesn't look like the type class is helping us that much.

<details> 
<summary>Maybe we should just keep it simple and ditch the polymorphism:
</summary>

```haskell
mkTmplFdWithDstAndData ::
  Path' (Rel WebAppTemplatesDir) File' ->
  Path' (Rel WebAppRootDir) File' ->
  Maybe Aeson.Value ->
  FileDraft
mkTmplFdWithDstAndData relSrcPath = 
  mkAnyValidTmplFdWithDstAndData (webAppTemplatesDirInTemplatesDir </> relSrcPath)

mkSharedTmplFdWithDst :: Path' (Rel SharedTemplatesDir) File' -> Path' (Rel WebAppRootDir) File' -> FileDraft
mkSharedTmplFdWithDst relSrcPath relDstPath =
  mkAnyValidTmplFdWithDstAndData
    (sharedTemplatesDirInTemplatesDir </> relSrcPath)
    relDstPath
    Nothing
    
mkAnyValidTmplFdWithDstAndData ::
  Path' (Rel TemplatesDir) File' ->
  Path' (Rel WebAppRootDir) File' ->
  Maybe Aeson.Value ->
  FileDraft
mkAnyValidTmplFdWithDstAndData relSrcPath relDstPath tmplData =
  createTemplateFileDraft
    (webAppRootDirInProjectRootDir </> relDstPath)
    relSrcPath
    tmplData
```

</details>

<details> 


> **Warning**
> I strongly prefer the previous option and am 85% joking with this one :)

<summary>Or, of course, we could double down:
</summary>

```haskell
class ValidWebAppTemplatesDir d where
  pathInTemplatesDir :: Path' (Rel TemplatesDir) (Dir d)

instance ValidWebAppTemplatesDir WebAppTemplatesDir where
  pathInTemplatesDir = webAppTemplatesDirInTemplatesDir

instance ValidWebAppTemplatesDir SharedTemplatesDir where
  pathInTemplatesDir = sharedTemplatesDirInTemplatesDir
  
-- All higher level functions would now be polymorphic and work automatically:

mkTmplFdWithDst :: ValidWebAppTemplatesDir d => Path' (Rel d) File' -> Path' (Rel WebAppRootDir) File' -> FileDraft
mkTmplFdWithDst src dst = mkTmplFdWithDstAndData src dst Nothing

mkTmplFdWithData :: ValidWebAppTemplatesDir d => Path' (Rel d) File' -> Aeson.Value -> FileDraft
mkTmplFdWithData src tmplData = mkTmplFdWithDstAndData src (SP.castRel src) (Just tmplData)

mkTmplFdWithDstAndData ::
  ValidWebAppTemplatesDir d =>
  Path' (Rel d) File' ->
  Path' (Rel WebAppRootDir) File' ->
  Maybe Aeson.Value ->
  FileDraft
mkTmplFdWithDstAndData relSrcPath relDstPath tmplData =
  createTemplateFileDraft
    (webAppRootDirInProjectRootDir </> relDstPath)
    (pathInTemplatesDir </> relSrcPath)
    tmplData
```

But we would then sometimes have to pay the verbosity tax when calling these functions (by specifying types for all arguments, one way or the other). For example:
```haskell
genEnvValidationScript :: Generator [FileDraft]
genEnvValidationScript =
  return
    [ C.mkTmplFd [relfile|scripts/validate-env.mjs|],
      C.mkTmplFdWithDst [relfile|validators.js|] [relfile|scripts/shared/validators.mjs|]
    ]
```
Would have to become:

```haskell
genEnvValidationScript :: Generator [FileDraft]
genEnvValidationScript =
  return
    [ C.mkTmplFd ([relfile|scripts/validate-env.mjs|] :: Path' (Rel WebAppTemplatesDir) File'),
      C.mkTmplFdWithDst
        ([relfile|validators.js|] :: Path' (Rel WebAppTemplatesDir) File')
        [relfile|scripts/shared/validators.mjs|]
    ]
```

</details>

With all that said, if you're tired of making changes, keeping what you currently have is fine too.

Reply:
I've went with the route of ditching polymorphism, keeping this simple.

We just introduced a `SharedTemplatesDir` concept and created a function for crafting `FileDrafts` from shared templates in `WebApp` and `Server` contexts. We expand on that as we use `shared` templates more.